File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/MainActivity.java
package com.shevaalex.android.rickmortydatabase.ui;

import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.NavigationUI;

import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.shevaalex.android.rickmortydatabase.R;
import com.shevaalex.android.rickmortydatabase.databinding.ActivityMainBinding;
import com.shevaalex.android.rickmortydatabase.ui.character.CharacterViewModel;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private static ArrayList<String> snackMessages = new ArrayList<>();
    private ActivityMainBinding binding;
    private NavController navController;
    private boolean backPressedOnce;
    private BottomNavViewModel botNavViewModel;
    private CharacterViewModel characterViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        botNavViewModel = new ViewModelProvider(this).get(BottomNavViewModel.class);
        characterViewModel = new ViewModelProvider.AndroidViewModelFactory(this.getApplication()).create(CharacterViewModel.class);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);
        setupViews();
        monitorConnectionAndDatabase();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
    }

    private void setupViews() {
        // Finding the navigation controller
        navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        // Setting the nav controller with bottom navigation
        NavigationUI.setupWithNavController(binding.bottomPanel, navController);
        //setup the ViewModel for lifecycle aware observing bottomNav state
        botNavViewModel.getBottomNavVisibility().observe(this, integer -> binding.bottomPanel.setVisibility(integer));
        botNavViewModel.getBottomNavLabelStatus().observe(this, integer -> binding.bottomPanel.setLabelVisibilityMode(integer));
        // monitor navigation and remove BottomNavigationView in Detail fragments
        navController.addOnDestinationChangedListener((controller, destination, arguments) -> {
            if (destination.getId() == R.id.settingsFragment || destination.getId() == R.id.characterImageFragment) {
                new Handler().postDelayed(() ->
                        botNavViewModel.hideBottomNav(), 100);
            } else if(destination.getId() == R.id.characterDetailFragment2 || destination.getId() == R.id.locationDetailFragment
                    || destination.getId() == R.id.episodeDetailFragment) {
                botNavViewModel.showBottomNav();
                botNavViewModel.setUnlabeled();
            } else {
                botNavViewModel.setLabelSelected();
                botNavViewModel.showBottomNav();
            }
        });
        // add bottom menu listener to prevent posibbility of double clicking the same item and refreshing or backing up the old search
        binding.bottomPanel.setOnNavigationItemSelectedListener(item -> {
            if (navController.getCurrentDestination() != null && navController.getCurrentDestination().getId() != item.getItemId()){
                NavigationUI.onNavDestinationSelected(item, navController);
                return true;
            }
            return true;
        });
    }

    //monitors internet connection, checks if database is up to date
    private void monitorConnectionAndDatabase() {
        characterViewModel.getStatusLiveData().observe(this, pair -> {
            @NonNull String text;
            int snackBarDuration = BaseTransientBottomBar.LENGTH_SHORT;
            // database is up to date and device is connected to network
            if (pair.first && pair.second) {
                binding.progressBar.progressBar.setVisibility(View.GONE);
                text = getString(R.string.ma_snack_database_up_to_date);
            }
            // database is _not_ up to date and device is connected to network
            else if (!pair.first && pair.second) {
                binding.progressBar.progressBar.setVisibility(View.VISIBLE);
                characterViewModel.rmRepository.initialiseDataBase();
                text = getString(R.string.ma_snack_database_sync);
            }
            // database is up to date and device is _disconnected_ from network
            else if (pair.first) {
                binding.progressBar.progressBar.setVisibility(View.VISIBLE);
                text = getString(R.string.ma_snack_database_up_to_date);
            }
            // database is _not_ up to date and device is _disconnected_ from network
            else {
                binding.progressBar.progressBar.setVisibility(View.VISIBLE);
                snackBarDuration = BaseTransientBottomBar.LENGTH_INDEFINITE;
                text = getString(R.string.ma_snack_database_not_synced);
            }
            showSnackBar(text, snackBarDuration);
        });
    }

    private void showSnackBar(String text, int snackBarDuration) {
        if (!text.isEmpty() && !snackMessages.contains(text)) {
            Snackbar mySnackbar = Snackbar.make(binding.activityMainLayout, text, snackBarDuration);
            mySnackbar.setTextColor(getResources().getColor(R.color.rm_white_50));
            mySnackbar.setAnchorView(binding.bottomPanel);
            mySnackbar.show();
            if (!text.equals(getString(R.string.ma_snack_database_not_synced))) {
                snackMessages.add(text);
            }
        }
    }

    @Override
    public void onBackPressed() {
        if(navController == null) {
            navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        }
        // Check if the current destination is actually the start destination (Home screen)
        if (navController.getCurrentDestination() != null
                && navController.getGraph().getStartDestination() == navController.getCurrentDestination().getId()){
            if(backPressedOnce){
                super.onBackPressed();
                return;
            }
            backPressedOnce = true;
            Toast.makeText(this, "Tap BACK again to exit", Toast.LENGTH_SHORT).show();
            new Handler().postDelayed(() -> backPressedOnce = false, 2000);
        } else { super.onBackPressed(); }
    }

}

File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/MainActivity.kt
package com.shevaalex.android.rickmortydatabase.ui

import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.Navigation
import androidx.navigation.findNavController
import androidx.navigation.ui.setupWithNavController
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.shevaalex.android.rickmortydatabase.R
import com.shevaalex.android.rickmortydatabase.RmApplication
import com.shevaalex.android.rickmortydatabase.databinding.ActivityMainBinding
import com.shevaalex.android.rickmortydatabase.ui.viewmodel.BottomNavViewModel
import com.shevaalex.android.rickmortydatabase.ui.viewmodel.InitViewModel
import com.shevaalex.android.rickmortydatabase.ui.viewmodel.ReviewViewModel
import com.shevaalex.android.rickmortydatabase.utils.DiViewModelFactory
import com.shevaalex.android.rickmortydatabase.utils.networking.Message
import com.shevaalex.android.rickmortydatabase.utils.networking.StateResource
import com.shevaalex.android.rickmortydatabase.utils.networking.Status
import kotlinx.coroutines.ExperimentalCoroutinesApi
import java.util.*
import javax.inject.Inject
import kotlin.concurrent.schedule

@ExperimentalCoroutinesApi
class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var viewModelFactory: DiViewModelFactory<InitViewModel>

    @Inject
    lateinit var reviewViewmodelFactory: DiViewModelFactory<ReviewViewModel>

    private lateinit var binding: ActivityMainBinding
    private var navController: NavController? = null
    private var backPressedOnce = false

    private val botNavViewModel: BottomNavViewModel by viewModels()
    private val initViewModel: InitViewModel by viewModels {
        viewModelFactory
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        (application as RmApplication).appComponent.inject(this)
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        setupNavController()
        registerObservers()
        setupEdgeToEdge()
    }

    private fun registerObservers() {
        //initialise or update the local database
        dbInit()
        //observe and set BottomNavigationView state
        botNavViewModel.bottomNavVisibility.observe(this,
            { integer: Int -> binding.bottomPanel.visibility = integer })
        botNavViewModel.bottomNavLabelVisibility.observe(this,
            { integer: Int -> binding.bottomPanel.labelVisibilityMode = integer })
    }

    private fun dbInit() {
        initViewModel.init().observe(this, {
            it?.let { stateResource ->
                val snackColor: Int?
                when (stateResource.status) {
                    is Status.Error -> {
                        snackColor = ContextCompat.getColor(this, R.color.rm_red_add)
                        composeMessage(stateResource, snackColor)
                        binding.progressBar.progressBar.visibility = View.GONE
                    }
                    is Status.Loading -> {
                        composeMessage(stateResource)
                        binding.progressBar.progressBar.visibility = View.VISIBLE
                    }
                    is Status.Success -> {
                        snackColor = ContextCompat.getColor(this, R.color.rm_green_300)
                        composeMessage(stateResource, snackColor)
                        binding.progressBar.progressBar.visibility = View.GONE
                        //notify viewmodel Db sync success
                        initViewModel.notifyDbAllSuccess()
                        unSubscribe()
                    }
                }
            }
        })
    }

    private fun unSubscribe() {
        initViewModel.init().removeObservers(this)
    }

    private fun setupEdgeToEdge() {
        //set window to draw behind system bars
        WindowCompat.setDecorFitsSystemWindows(window, false)
        //update the padding of the BottomNavigationView
        binding.bottomPanel.setOnApplyWindowInsetsListener { view, insets ->
            val insetsCompat = WindowInsetsCompat.toWindowInsetsCompat(insets)
            val systemWindow = insetsCompat.getInsets(
                WindowInsetsCompat.Type.navigationBars() or WindowInsetsCompat.Type.ime()
            )
            view.updatePadding(bottom = systemWindow.bottom)
            insets
        }
    }

    private fun setupNavController() {
        val navController = findNavController(R.id.nav_host_fragment)
        binding.bottomPanel.setupWithNavController(navController)
        // monitor navigation and set unlabeled BottomNavigationView in Detail fragments
        navController.addOnDestinationChangedListener { _, destination: NavDestination, _ ->
            if (destination.id == R.id.settingsFragment) {
                botNavViewModel.hideBottomNav()
            } else if (destination.id == R.id.characterDetailFragment2
                || destination.id == R.id.locationDetailFragment
                || destination.id == R.id.episodeDetailFragment
            ) {
                botNavViewModel.showBottomNav()
                botNavViewModel.setUnlabeled()
            } else {
                botNavViewModel.showBottomNav()
                botNavViewModel.setLabelSelected()
            }
        }
    }

    private fun composeMessage(stateResource: StateResource, snackColor: Int? = null) {
        val snackText: String
        when (stateResource.message) {
            is Message.NoInternet -> {
                snackText = getString(R.string.ma_snack_database_not_synced)
            }
            is Message.DbIsUpToDate ->
                snackText = getString(R.string.ma_snack_database_up_to_date)
            is Message.ServerError ->
                snackText = getString(R.string.ma_snack_error_server_error)
                    .plus(stateResource.message.statusCode)
            is Message.NetworkError ->
                snackText = getString(R.string.ma_snack_error_network_error)
            is Message.EmptyResponse ->
                snackText = getString(R.string.ma_snack_error_empty_response)
            null ->
                snackText = ""
        }
        if (snackText.isNotBlank()) {
            showSnackBar(snackText, snackColor)
        }
    }

    private fun showSnackBar(text: String, color: Int? = null) {
        val mySnackbar = Snackbar
            .make(binding.activityMainLayout, text, BaseTransientBottomBar.LENGTH_LONG)
        val snackBarView = mySnackbar.view
        color?.let {
            snackBarView.rootView.setBackgroundColor(it)
        }
        mySnackbar.setTextColor(ContextCompat.getColor(this, R.color.rm_white_50))
        mySnackbar.anchorView = binding.bottomPanel
        mySnackbar.show()
    }

    override fun onBackPressed() {
        if (navController == null) {
            navController = Navigation.findNavController(this, R.id.nav_host_fragment)
        }
        // Check if the current destination is actually the start destination (Home screen)
        if (navController?.currentDestination != null
            && navController?.graph?.startDestination == navController?.currentDestination?.id
        ) {
            if (backPressedOnce) {
                super.onBackPressed()
                return
            }
            backPressedOnce = true
            showSnackBar(resources.getString(R.string.toast_close_message))
            Timer().schedule(2000) {
                backPressedOnce = false
            }
        } else {
            super.onBackPressed()
        }
    }

}